streaming/base/dataset.py

```diff
@@ -366,14 +366,53 @@ def _get_partition(self, world: World, epoch: int, sample_in_epoch: int) -> NDAr
         if self.shuffle_seed is None:
             raise RuntimeError('Shuffle seed can never be None')
 
-        sample_ids = get_partitions(self.index.total_samples, self.num_canonical_nodes,
-                                    world.num_nodes, world.ranks_per_node, world.workers_per_rank,
-                                    self.batch_size, sample_in_epoch)
-        if self.shuffle:
-            mapping = get_shuffle(self.shard_sizes, self.num_canonical_nodes, self.shuffle_seed,
-                                  epoch)
-            sample_ids = np.where(sample_ids == -1, -1, mapping[sample_ids])
-        return sample_ids[world.node, world.rank_of_node, world.worker_of_rank]
+        tmp_filename = os.path.join(os.path.sep, 'tmp', 'streaming', self._prefix,
+                                    'shuffle.npy.tmp')
+        filename = os.path.join(os.path.sep, 'tmp', 'streaming', self._prefix, 'shuffle.npy')
+
+        # In the local leader, generate this epoch's global sample ordering, then save to file.
+        # This operation is expensive.
+        if world.is_local_leader:
+            sample_ids = get_partitions(self.index.total_samples, self.num_canonical_nodes,
+                                        world.num_nodes, world.ranks_per_node,
+                                        world.workers_per_rank, self.batch_size, sample_in_epoch)
+            if self.shuffle:
+                mapping = get_shuffle(self.shard_sizes, self.num_canonical_nodes,
+                                      self.shuffle_seed, epoch)
+                sample_ids = np.where(sample_ids == -1, -1, mapping[sample_ids])
+            sample_ids.tofile(tmp_filename)
+            os.rename(tmp_filename, filename)
+
+        # Everyone waits for the file to become populated.
```

Comment:
Should we add any exit strategy that would fail after couple of mins or an hour to avoid indefinite hang in `while True` loop ?

Reply:
done